### PR TITLE
[RepairManager] MonitorSMILatencyRule

### PR DESCRIPTION
**Pull request by debbie-alaine**
added MonitorSMILatencyRule:
- monitors prometheus for large latency on node 
  - query="histogram_quantile(0.95, sum(rate(cmd_nvidia_smi_latency_seconds_bucket[5m])) BY (le, instance)) > 40"
- if found, send alert to DRI
- alert expires in 4 hours (configurable).
  - i.e. if issue does not resolve on prometheus, alert will be resent

image

Other things:
- refactored some code into prometheus_util
- refactored some code into k8s_util
- refactored some code into test_util
- added some unit tests for MonitorSMILatencyRule

**Testing in Dev1 Complete**

### CODE REVIEW

File: src/ClusterBootstrap/template/RepairManager/ecc-config.yaml [hunk 1]
```diff
@@ -2,13 +2,6 @@ enable_cordon: {{cnf['repair-manager']['ecc_rule']['enable_cordon']}}
 enable_reboot: {{cnf['repair-manager']['ecc_rule']['enable_reboot']}}
 enable_alert_job_owners: {{cnf['repair-manager']['ecc_rule']['enable_alert_job_owners']}}
 
-prometheus:
-  ip: {{cnf['repair-manager']['prometheus-ip']}}
-  port: {{cnf['repair-manager']['prometheus-port']}}
-  ecc_error_query: 'nvidiasmi_ecc_error_count{type="volatile_double"}>0'
-  node_boot_time_query: 'node_boot_time_seconds'
-
-rest_url: {{cnf['repair-manager']['rest_url']}}
```
Comment (debbie-alaine) [inline src/ClusterBootstrap/template/RepairManager/ecc-config.yaml]: moved this out to rule-config.yaml, to be shared with other rules.

File: src/RepairManager/rules/ecc_detect_error_rule.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 from actions.cordon_action import CordonAction
 from actions.send_alert_action import SendAlertAction
 from kubernetes import client, config
-from utils import k8s_util, email_util, prometheus_url
+from utils import k8s_util, email_util, prometheus_util
```
Comment (debbie-alaine) [inline src/RepairManager/rules/ecc_detect_error_rule.py]: uniformity in naming convention

File: src/RepairManager/rules/monitor_smi_latency_rule.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                time_found_string = self.alert.rule_cache[self.rule][node]["time_found"]
+                time_found_datetime = datetime.strptime(time_found_string, self.config['date_time_format'])
+                alert_delta = timedelta(days=self.latency_config.get("hours_until_alert_expiration", 4))
+                now = datetime.utcnow()
+                if now - time_found_datetime > alert_delta:
+                    self.alert.remove_from_rule_cache()
+
+
+    def check_status(self):
+        self.clean_expired_items_in_rule_cache()
```
Comment (debbie-alaine) [inline src/RepairManager/rules/monitor_smi_latency_rule.py]: Added this to clean out the expired entries from rule cache before executing logic.

Open to other suggestions, a TTLCache maybe?

File: src/ClusterBootstrap/template/RepairManager/rule-config.yaml [hunk 4]
```diff
@@ -6,6 +6,9 @@ rules:
   ecc_reboot_node_rule:
     module_name: rules.ecc_reboot_node_rule
     class_name: ECCRebootNodeRule
+  monitor_smi_latency_rule:
+    module_name: rules.monitor_smi_latency_rule
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/template/RepairManager/rule-config.yaml]: Can we use `nvidia_smi_latency_rule` for clarity? The corresponding class name would be `NvidiaSmiLatencyRule`

File: src/RepairManager/test/test_ecc_detect_error_rule.py [hunk 5]
```diff
@@ -7,73 +7,10 @@
...
-                "success",
-                "data": {
-                    "result": []
-                }
-        }
-    return empty_prometheus_error_data
-
-def _mock_prometheus_error_data():
-    mock_prometheus_error_data = {
+def _mock_prometheus_latency_data():
```
Comment (Anbang-Hu) [inline src/RepairManager/test/test_ecc_detect_error_rule.py]: I am pretty sure this is not latency data.

File: src/RepairManager/test/test_ecc_detect_error_rule.py [hunk 6]
```diff
@@ -107,54 +44,14 @@ def _mock_prometheus_error_data():
                     }]
                 }
         }
-    return mock_prometheus_error_data
+    return mock_prometheus_latency_data
 
 class Testing(unittest.TestCase):
```
Comment (Anbang-Hu) [inline src/RepairManager/test/test_ecc_detect_error_rule.py]: Usually the test class for a class or a file is named `Test<CamelCaseClassOrFileName>`, hence `TestEccDetectErrorRule`

File: src/ClusterBootstrap/template/RepairManager/latency-config.yaml [hunk 7]
```diff
@@ -0,0 +1 @@
+hours_until_alert_expiration: {{cnf['repair-manager']['latency_rule']['hours_until_alert_expiration']}}
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/template/RepairManager/latency-config.yaml]: This is a pretty long name. `alert_expiry`? Hashtag comment it's in hours should be sufficient.

File: src/RepairManager/rules/monitor_smi_latency_rule.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        for node in expired_nodes:
+            self.alert.remove_from_rule_cache(self.rule, node)
+
+
+    def check_status(self):
+        self.clean_expired_items_in_rule_cache()
+
+        url = f"http://{self.config['prometheus']['ip']}:{self.config['prometheus']['port']}"
+        query = self.config['prometheus']['large_latency_query']
+        ecc_url = prometheus_util.format_url_query(url, query)
```
Comment (Anbang-Hu) [inline src/RepairManager/rules/monitor_smi_latency_rule.py]: Not `ecc_url`. Please change the names in line 83, 97, 99 accordingly.

File: src/RepairManager/rules/nvidia_smi_latency_rule.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                node_ips = prometheus_util.extract_ips_from_response(response)
+                if node_ips:
+                    address_map = k8s_util.get_node_address_info()
+                    for ip in node_ips:
+                        node_name = address_map[ip]
+                        if not self.alert.check_rule_cache(self.rule, node_name):
+                            self.impacted_nodes[node_name] = ip
+                    return len(self.impacted_nodes) > 0
+                else:
+                    logging.debug('No uncorrectable ECC metrics found.')
```
Comment (Anbang-Hu) [inline src/RepairManager/rules/nvidia_smi_latency_rule.py]: The debug message doesn't look right

**Pull request merged**